text/0062-controlling-pii-and-credentials-in-sd-ks.md

```diff
@@ -0,0 +1,207 @@
+
+As demonstrated by the collection above, each SDK does slightly different things and the decisions on what is gated behind `send_default_pii` are mostly left to the integration authors.  
+The user can merely choose to follow the author's implementation with this flag but not really be sure if and what data is leaking without reading the integration source, which is an undesirable state of affairs.
+
+
+#### `before_send` and `add_event_processor`
+
+In cases where users are not happy with `send_default_pii`, we suggest using either `before_send` for errors or `add_event_processor` for all events for controlling PII on the final event payload.  
+This, while it works in theory has several problems:
+
+* there is an ordering problem when a user's added global event processor runs before some other event processor added by an integration, thus not giving the user the actual final event
+  * there are [documented **unsupported** use cases](https://github.com/getsentry/sentry-python/issues/1226#issuecomment-978806984) here such as the user never being able to remove a certain request header since the headers are actually added [after](https://github.com/getsentry/sentry-python/blob/de84c4c7d65830e91f3af8555ebb1f7385a2d7cc/sentry_sdk/integrations/wsgi.py#L306-L325) the user's event processor runs
+  * this will be partially fixed by `before_send_transaction` but I still believe PII scrubbing is a key concern that should not be shoe-horned into catch-all callbacks that run at the end
+* writing either of these callbacks requires the user to know the whole event schema which even our own SDK devs don't know fully
+
+# Options Considered
+
+The high-level proposal is to introduce a new `config.event_scrubber` with a default implementation that users can tweak with a simple `denylist` in most cases.  
+The end-user only cares about hiding certain information quickly, so a single list is the ideal interface.  
+Anything more complex than that is asking them to think about things they don't care about.  
+The advanced user can still override the implementation (similar to custom transports) if they wish to do so by using their own class implementation.
+
+
+```python
+# ===========================================================================
+# event_scrubber.py
+
+DEFAULT_DENYLIST = ["password", "HTTP_COOKIE"]
+
+class EventScrubber:
+    def __init__(self, denylist=None):
+        self.denylist = DEFAULT_DENYLIST if denylist is None else denylist
+
+    def scrub_event(self, event):
+        # type: (Event) -> Event
+        # ... scrubbing logic discussed below
+        return event
+
+
+# ===========================================================================
+# client.py
+
+def capture_event(self, event):
```

Comment:
So here we would expect an SDK to process an event like event processors -> event scrubber -> before send?

Reply:
yep, since we use event processors internally for integrations as well, I would say so.